src/librustc_mir/interpret/const_eval.rs

```diff
@@ -570,3 +590,40 @@ pub fn const_eval_provider<'a, 'tcx>(
         }
     })
 }
+
+fn numeric_intrinsic<'tcx>(
+    name: &str,
+    bytes: u128,
+    kind: Primitive,
+) -> EvalResult<'tcx, Scalar> {
+    macro_rules! integer_intrinsic {
+        ($method:ident) => ({
+            use rustc_target::abi::Integer;
+            let (bits, defined) = match kind {
+                Primitive::Int(Integer::I8, true) => ((bytes as i8).$method() as u128, 8),
```

Comment:
@eddyb I know you dislike these kinds of matches. We could probably do all the ops on u128 and modify the result depending on the type's bitwidth, but that feels fragile

Reply:
`let extra = 128 - bits;`
* `ctpop(x) = x.count_ones()`
* `ctlz(x) = x.leading_zeros() - extra`
* `cttz(x) = (x << extra).trailing_zeros() - extra`
* `bswap(x) = (x << extra).swap_bytes()`

Reply:
I agree the code can probably be written a bit nicer. Looking into it. Will push as soon as I have the time.